src/Workspaces/Core/Portable/EmbeddedLanguages/VirtualChars/VirtualCharSequence.cs

```diff
@@ -0,0 +1,148 @@
+    ///
+    /// Then you'd get back:
+    ///
+    /// 'G' -> [0, 1) 'o' -> [1, 7) 'o' -> [7, 1)
+    ///
+    /// This allows for embedded language processing that can refer back to the user's original code
+    /// instead of the escaped value we're processing.
+    /// </summary>
+    internal partial struct VirtualCharSequence
+    {
+        public static readonly VirtualCharSequence Empty = Create(ImmutableArray<VirtualChar>.Empty);
+
+        public static VirtualCharSequence Create(ImmutableArray<VirtualChar> virtualChars)
+            => new VirtualCharSequence(new ImmutableArrayChunk(virtualChars));
+
+        public static VirtualCharSequence Create(int firstVirtualCharPosition, string underlyingData)
+            => new VirtualCharSequence(new StringChunk(firstVirtualCharPosition, underlyingData));
+
+        /// <summary>
+        /// The actual characters that this <see cref="VirtualCharSequence"/> is a portion of.
+        /// </summary>
+        private readonly Chunk _leafCharacters;
+
+        /// <summary>
+        /// The portion of <see cref="_leafCharacters"/> that is being exposed.  This span 
+        /// is `[inclusive, exclusive)`.
+        /// </summary>
+        private readonly TextSpan _span;
+
+        private VirtualCharSequence(Chunk sequence) : this(sequence, new TextSpan(0, sequence.Length))
+        {
+        }
+
+        private VirtualCharSequence(Chunk sequence, TextSpan span)
+        {
+            if (span.Start > sequence.Length)
+            {
+                throw new ArgumentException();
+            }
+
+            if (span.End > sequence.Length)
+            {
+                throw new ArgumentException();
+            }
+
+            _leafCharacters = sequence;
+            _span = span;
+        }
+
+        public int Length => _span.Length;
+        public VirtualChar this[int index] => _leafCharacters[_span.Start + index];
+
+        public bool IsDefault => _leafCharacters == null;
+        public bool IsEmpty => Length == 0;
+        public bool IsDefaultOrEmpty => IsDefault || IsEmpty;
+
+        public VirtualCharSequence GetSubSequence(TextSpan span)
+           => new VirtualCharSequence(
+               _leafCharacters, new TextSpan(_span.Start + span.Start, span.Length));
+
+        public VirtualChar First() => this[0];
+        public VirtualChar Last() => this[this.Length - 1];
+
+        public Enumerator GetEnumerator()
+            => new Enumerator(this);
+
+        public VirtualChar? FirstOrNullable(Func<VirtualChar, bool> predicate)
+        {
+            foreach (var ch in this)
+            {
+                if (predicate(ch))
+                {
+                    return ch;
+                }
+            }
+
+            return null;
+        }
+
+        public bool Contains(VirtualChar @char)
+            => IndexOf(@char) >= 0;
+
+        public int IndexOf(VirtualChar @char)
+        {
+            int index = 0;
+            foreach (var ch in this)
```

Comment:
nit: why do a `foreach` when you already have an index? `for` seems appropriate here

Reply:
`foreach` can be more efficient *if* the underlying representation changes.  For example, if we cahnge this to a tree in the future.  However, indexin may end up being `O(n log n)`.  This style keeps us safe.